src/stpipe/library.py

```diff
@@ -0,0 +1,850 @@
+                    break
+                if isinstance(model_or_filename, (str, Path)):
+                    # Supporting a list of filenames by opening them as models
+                    # has issues, if this is a widely supported mode (vs providing
+                    # an association) it might make the most sense to make a fake
+                    # association with the filenames at load time.
+                    model = self._datamodels_open(
+                        model_or_filename, **self._datamodels_open_kwargs
+                    )
+                else:
+                    model = model_or_filename
+
+                exptype = getattr(model.meta, "exptype", "SCIENCE")
+
+                if asn_exptypes is not None and exptype.lower() not in asn_exptypes:
+                    continue
+
+                members.append(
+                    {
+                        "expname": self._model_to_filename(model),
+                        "exptype": exptype,
+                        "group_id": self._to_group_id(model, len(members)),
+                    }
+                )
+
+                self._loaded_models[len(self._loaded_models)] = model
+
+            # since we've already filtered by asn type and n members reset these values
+            asn_exptypes = None
+            asn_n_members = None
+
+            # make a very limited fake association
+            asn_data = {
+                "products": [
+                    {
+                        "members": members,
+                    }
+                ],
+            }
+            self._asn_dir = None
+        else:
+            raise ValueError(f"Invalid init {init}")
+
+        if asn_exptypes is not None:
+            asn_data["products"][0]["members"] = [
+                m
+                for m in asn_data["products"][0]["members"]
+                if m["exptype"] in asn_exptypes
+            ]
+
+        if asn_n_members is not None:
+            asn_data["products"][0]["members"] = asn_data["products"][0]["members"][
+                :asn_n_members
+            ]
+
+        self._asn = asn_data
+        self._members = self._asn["products"][0]["members"]
+
+        for index, member in enumerate(self._members):
+            if "group_id" not in member:
+                filename = os.path.join(self._asn_dir, member["expname"])
+                member["group_id"] = self._to_group_id(filename, index)
+
+        if not on_disk:
+            # if models were provided as input, assign the members here
+            # now that the fake asn data is complete
+            for i in self._loaded_models:
+                self._assign_member_to_model(self._loaded_models[i], self._members[i])
+
+    def __del__(self):
+        if hasattr(self, "_temp_dir"):
+            self._temp_dir.cleanup()
+
+    @property
+    def asn(self):
+        """
+        Association dictionary used to create the library.
+
+        Note that changes to the models may cause this information
```

Comment:
Good question. I'd say yes since the library will not use the new `group_id` (it will be "saved" in the model but any grouping will use the `group_id` at the time of creation). Also the `library.asn` will not list the updated `group_id`.